File: Atividade4/v1/alggenetico.m


function AG_Func = alggenetico(MatrizDistTrab,tam_Pop_ini,tam_gera)

tempoAG = tic;
DistMatriz = MatrizDistTrab;

numCidades = length(DistMatriz);

%   Numero de Filhos
%nmr_filhos = tam_Pop_ini - 2;
%palim = nmr_filhos/2;
%malim = nmr_filhos/2;

%   Alocação de variaveis para aumentar a velocidade
Populacao = zeros(tam_Pop_ini, (numCidades+1));
FilhosPai = zeros(4, (numCidades+1));
FilhosMae = zeros(4, (numCidades+1));

% Gere a população inicial contendo todas as cidades, mas em ordem aleatória
Percurso = zeros(tam_Pop_ini, numCidades+1); % +1 para incluir a cidade de origem no fim da rota

%Geracao da Populacao Inicial
for k = 1:tam_Pop_ini
    
    % Gera um vetor de cidades (exceto a cidade de origem) em ordem aleatória
    cidades_aleatorias = randperm(numCidades-1) + 1;
    
    % Cria o cromossomo (rota) com a cidade de origem no início e fim
    Populacao(k, :) = [1, cidades_aleatorias, 1];


end


for g = 1:tam_gera
    

    %Calcula o percurso total de cada vetor solucao
    for k = 1:size(Populacao,1)
        Percurso(k, g) = Calc_Dist(DistMatriz, Populacao(k,:));
    end

    %Escolhe os mais aptos 
    clear a b;
    [a, b] = sort(Percurso(:,g));
    
    % Recupera a menor distância encontrada
    menorDistancia = a(1);

    % Recupera a melhor rota encontrada
    melhorRota = Populacao(b(1,1), :);

    Pai = Populacao(b(1,1), :);
    Mae = Populacao(b(2,1), :);

    %Cada Genitor vai gerar 2 filhos por permuta
    %Sorteia e permuta as duas posicoes do Pai
    for pa = 1:4
        clear a b;
        [~, b] = sort(rand(2,numCidades));
        %b = b + 1;

        FilhosPai(pa, :) = Pai;
        FilhosPai(pa, b(1, 1)) = Pai(1, b(1, 2));
        FilhosPai(pa, b(1, 2)) = Pai(1, b(1, 1));
    end

    %Sorteia e permuta as duas posicoes da Mae
    for ma = 1:4
        clear a b;
        [~, b] = sort(rand(2,numCidades));
        %b = b + 1;

        FilhosMae(ma, :) = Mae;
        FilhosMae(ma, b(1, 1)) = Mae(1, b(1, 2));
        FilhosMae(ma, b(1, 2)) = Mae(1, b(1, 1));
    end


    %Atualiza a Populacao
    Populacao = [Pai; Mae; FilhosPai; FilhosMae];
  
end

tempoAG = toc(tempoAG);
%plot(min(Percurso, [], 1));


% Preparando a saída da função
AG_Func.tempoAG = tempoAG;
AG_Func.menorDistancia = menorDistancia;
AG_Func.melhorRota = melhorRota;







end

File: Python/alggenetico.py
"""Arquivo para metodo Algortimo Genetico de Selecao Simples"""
import numpy as np
from Calc_func import Calc_Dist
import time


""" Funcao para executar o algoritmo genetico """


def alggenetico(MatrizDistTrab, tam_Pop_ini, tam_gera, tx_mutacao):
    tempoAG = time.time()
    DistMatriz = MatrizDistTrab

    numCidades = len(DistMatriz)

    # Criando as variaveis para economizar memoria
    Populacao = np.zeros((tam_Pop_ini, (numCidades+1)))
    nmr_filhos = np.zeros(((tam_Pop_ini-2), (numCidades+1)))
    Percurso = np.zeros((tam_Pop_ini, tam_gera))

    # Criando a populacao inicial
    for k in range(tam_Pop_ini):
        cidades_aleatorias = np.random.choice(np.setdiff1d(
            np.arange(2, numCidades+1), [1]), size=numCidades-1, replace=False)

        Populacao[k, :] = np.concatenate(([0], cidades_aleatorias, [0]))

    # Calculando o percurso
    for g in range(tam_gera):
        for k in range(Populacao.shape[0]):
            Percurso[k, g] = Calc_Dist(DistMatriz, Populacao[k, :])

        # Escolhendo os mais aptos
        indices_ordenados = np.argsort(Percurso[:, g])
        menorDistancia = Percurso[indices_ordenados[0], g]
        melhorRota = Populacao[indices_ordenados[0], :]

        # Selecionando os pais
        Pai = Populacao[indices_ordenados[0], :].reshape(1, -1)
        Mae = Populacao[indices_ordenados[1], :].reshape(1, -1)

        # Definindo os filhos
        num_filhos = nmr_filhos.shape[0]

        if num_filhos % 2 == 0:
            palim = num_filhos // 2
            malim = num_filhos // 2
        else:
            escolha = np.random.randint(1, 3)
            if escolha == 1:
                palim = (num_filhos + 1) // 2
                malim = (num_filhos - 1) // 2
            else:
                palim = (num_filhos - 1) // 2
                malim = (num_filhos + 1) // 2

        # Criando os filhos
        FilhosPai = np.zeros((palim, (numCidades+1)))
        FilhosMae = np.zeros((malim, (numCidades+1)))

        # Criando os filhos por permutacao e aplicando mutacao
        for pa in range(palim):
            # Selecionado os indices para a troca de cidades
            b = np.random.permutation(np.arange(1, numCidades))[:2]
            FilhosPai[pa, :] = Pai
            FilhosPai[pa, b[0]] = Pai[0, b[1]]
            FilhosPai[pa, b[1]] = Pai[0, b[0]]

            # Aplicando a Mutacao com base na probabilidade
            if np.random.rand() < tx_mutacao:
                m = np.random.permutation(np.arange(1, numCidades))[:2]
                FilhosPai[pa, :] = Pai
                FilhosPai[pa, m[0]] = Pai[0, m[1]]
                FilhosPai[pa, m[1]] = Pai[0, m[0]]

        for ma in range(malim):
            # Selecionado os indices para a troca de cidades
            b = np.random.permutation(np.arange(1, numCidades))[:2]
            FilhosMae[ma, :] = Mae
            FilhosMae[ma, b[0]] = Mae[0, b[1]]
            FilhosMae[ma, b[1]] = Mae[0, b[0]]

            # Aplicando a Mutacao com base na probabilidade
            if np.random.rand() < tx_mutacao:
                m = np.random.permutation(np.arange(1, numCidades))[:2]
                FilhosMae[ma, :] = Mae
                FilhosMae[ma, m[0]] = Mae[0, m[1]]
                FilhosMae[ma, m[1]] = Mae[0, m[0]]

        #   Formando a nova populacao
        Populacao = np.concatenate((Pai, Mae, FilhosPai, FilhosMae))

    tempoAG = time.time() - tempoAG

    AG_Func = {}
    AG_Func['tempoAG'] = tempoAG
    AG_Func['menorDistancia'] = menorDistancia
    AG_Func['melhorRota'] = melhorRota

    return AG_Func
